File: src/main/java/ru/org/linux/tag/TagModificationService.java
/*
 * Copyright 1998-2013 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.tag;

import com.google.common.base.Strings;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.validation.Errors;
import ru.org.linux.user.UserErrorException;
import scala.Option;

import java.util.ArrayList;
import java.util.List;

@Service
public class TagModificationService {
  private static final Logger logger = LoggerFactory.getLogger(TagModificationService.class);

  @Autowired
  private TagDao tagDao;

  @Autowired
  private TagService tagService;

  private final List<ITagActionHandler> actionHandlers = new ArrayList<>();

  public List<ITagActionHandler> getActionHandlers() {
    return actionHandlers;
  }

  /**
   * Изменить название существующего тега.
   *
   * @param oldTagName старое название тега
   * @param tagName    новое название тега
   * @param errors     обработчик ошибок ввода для формы
   */
  public void change(String oldTagName, String tagName, Errors errors) {
    try {
      TagName.checkTag(tagName);
      int oldTagId = tagService.getTagId(oldTagName);

      if (tagDao.getTagId(tagName).isDefined()) {
        errors.rejectValue("tagName", "", "Тег с таким именем уже существует!");
      } else {
        tagDao.changeTag(oldTagId, tagName);
        logger.info(
                "Изменено название тега. Старое значение: '{}'; новое значение: '{}'",
                oldTagName,
                tagName
        );
      }
    } catch (UserErrorException e) {
      errors.rejectValue("tagName", "", e.getMessage());
    } catch (TagNotFoundException e) {
      errors.rejectValue("tagName", "", "Тега с таким именем не существует!");
    }
  }

  /**
   * Удалить тег по названию. Заменить все использования удаляемого тега
   * новым тегом (если имя нового тега не null).
   *
   * @param tagName    название тега
   * @param newTagName новое название тега
   * @param errors     обработчик ошибок ввода для формы
   */
  @Transactional(rollbackFor = Exception.class, propagation = Propagation.REQUIRED)
  public void delete(String tagName, String newTagName, Errors errors) {
    try {
      int oldTagId = tagService.getTagId(tagName);
      if (!Strings.isNullOrEmpty(newTagName)) {
        if (newTagName.equals(tagName)) {
          errors.rejectValue("tagName", "", "Заменяемый тег не должен быть равен удаляемому!");
          return;
        }
        TagName.checkTag(newTagName);
        int newTagId = getOrCreateTag(newTagName);

        for (ITagActionHandler actionHandler : actionHandlers) {
          actionHandler.replaceTag(oldTagId, newTagId, newTagName);
        }
        logger.debug("Удаляемый тег '{}' заменён тегом '{}'", tagName, newTagName);
      }
      for (ITagActionHandler actionHandler : actionHandlers) {
        actionHandler.deleteTag(oldTagId, tagName);
      }
      tagDao.deleteTag(oldTagId);
      logger.info("Удалён тег: " + tagName);
    } catch (UserErrorException e) {
      errors.rejectValue("tagName", "", e.getMessage());
    } catch (TagNotFoundException e) {
      errors.rejectValue("tagName", "", "Тега с таким именем не существует!");
    }
  }

  /**
   * Получение идентификационного номера тега по названию, либо создание нового тега.
   *
   * @param tagName название тега
   * @return идентификационный номер тега
   */
  public int getOrCreateTag(final String tagName) {
    Option<Integer> tagId = tagDao.getTagId(tagName);

    if (tagId.isDefined()) {
      return tagId.get();
    } else {
      return tagDao.createTag(tagName);
    }
  }
}

File: src/main/scala/ru/org/linux/tag/TagModificationService.scala
/*
 * Copyright 1998-2023 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */
package ru.org.linux.tag

import com.typesafe.scalalogging.StrictLogging
import org.springframework.scala.transaction.support.TransactionManagement
import org.springframework.stereotype.Service
import org.springframework.transaction.PlatformTransactionManager
import ru.org.linux.search.SearchQueueSender
import ru.org.linux.topic.TopicTagDao
import ru.org.linux.user.UserTagDao

import scala.collection.mutable

@Service
class TagModificationService(val transactionManager: PlatformTransactionManager, tagDao: TagDao,
                             tagService: TagService, userTagDao: UserTagDao, topicTagDao: TopicTagDao,
                             searchQueueSender: SearchQueueSender) extends StrictLogging with TransactionManagement {
  /**
    * Изменить название существующего тега.
    *
    * @param oldTagName старое название тега
    * @param tagName    новое название тега
    */
  def change(oldTagName: String, tagName: String): Unit = transactional() { _ =>
    val oldTagId = tagService.getTagId(oldTagName)

    tagDao.deleteTagSynonym(tagName)
    tagDao.changeTag(oldTagId, tagName)

    topicTagDao.processTopicsByTag(oldTagId, searchQueueSender.updateMessage(_, true))
  }

  /**
    * Удалить тег по названию.
    *
    * @param tagName    название тега
    */
  def delete(tagName: String): Unit = {
    val toUpdate = transactional() { _ =>
      val oldTagId = tagService.getTagId(tagName)

      val toUpdate = mutable.Buffer[Int]()
      topicTagDao.processTopicsByTag(oldTagId, toUpdate += _)

      userTagDao.deleteTags(oldTagId)
      topicTagDao.deleteTag(oldTagId)

      tagDao.deleteTag(oldTagId)

      toUpdate
    }

    toUpdate.foreach(searchQueueSender.updateMessage(_, true))
  }

  /**
    * Удалить тег по названию. Заменить все использования удаляемого тега
    * новым тегом.
    *
    * @param tagName    название тега
    * @param newTagName новое название тега
    */
  def merge(tagName: String, newTagName: String, createSynonym: Boolean): Unit = {
    val newTagId = transactional() { _ =>
      val oldTagId = tagService.getTagId(tagName)

      assume(newTagName != tagName, "Заменяемый тег не должен быть равен удаляемому")

      val newTagId = tagService.getOrCreateTag(newTagName)

      val tagCount = topicTagDao.getCountReplacedTags(oldTagId, newTagId)
      topicTagDao.replaceTag(oldTagId, newTagId)
      topicTagDao.increaseCounterById(newTagId, tagCount)

      userTagDao.replaceTag(oldTagId, newTagId)
      tagDao.updateTagSynonym(oldTagId, newTagId)

      logger.info("Удаляемый тег '{}' заменён тегом '{}'", tagName, newTagName)

      userTagDao.deleteTags(oldTagId)
      topicTagDao.deleteTag(oldTagId)

      tagDao.deleteTag(oldTagId)

      if (createSynonym) {
        tagDao.createTagSynonym(tagName, newTagId)
      }

      newTagId
    }

    topicTagDao.processTopicsByTag(newTagId, searchQueueSender.updateMessage(_, true))
  }
}